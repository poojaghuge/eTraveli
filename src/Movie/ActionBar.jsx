import React, { useState } from "react";
import { styled, alpha } from "@mui/material/styles";
import {
  Paper,
  Box,
  Toolbar,
  InputBase,
  InputLabel,
  MenuItem,
  OutlinedInput,
} from "@mui/material";
import SearchIcon from "@mui/icons-material/Search";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import "./Movie.css";

const StyledSelect = styled(Select)(({ theme }) => ({
  color: "inherit",
  "& .MuiSelect-select, .MuiInputBase-input": {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `${theme.spacing(4)})`,
    [theme.breakpoints.up("md")]: {
      width: "10ch",
    },
  },
}));



function ActionBar(props) {
  const [selectedValue, setSelectedValue] = useState("");

  const sortBy = ["Year", "Episode"];
  const handleSortChange = (event) => {
    const {
      target: { value },
    } = event;
    setSelectedValue(value);
    const order = value == "Year" ? "release_date" : "episode_id";
    props.handleSort(order);
  };

  return (
    <Paper sx={{ flexGrow: 1 }} variant={"outlined"}>
      <Toolbar>
        <FormControl sx={{ m: 1 }} className="sortBy">
          <StyledSelect
            labelId="sort-label"
            id="sort"
            displayEmpty
            value={selectedValue}
            onChange={handleSortChange}
            renderValue={(selectedValue) => {
              return <em>Sort by...{selectedValue}</em>;
            }}>
            <MenuItem disabled value="">
              <em>Sort by...</em>
            </MenuItem>
            {sortBy.map((name) => (
              <MenuItem key={name} value={name}>
                {name}
              </MenuItem>
            ))}
          </StyledSelect>
        </FormControl>
        <div className="search">
          <SearchIcon className="searchIcon" />
          <InputBase
            className="searchInput"
            placeholder="Type to search..."
            inputProps={{ "aria-label": "search" }}
            onChange={(event) => props.handleChange(event.target.value)}
          />
        </div>
        <Box sx={{ flexGrow: 1 }} />
      </Toolbar>
    </Paper>
  );
}

export const MemoisedActionBar = React.memo(ActionBar);
