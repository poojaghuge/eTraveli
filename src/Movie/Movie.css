.search {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  border: 1px solid #f0f0f0e6;
  padding: 3px;
  border-radius: 5px;
}
.container{
    /* display: flex;
    justify-content: center;
    align-items: center; */
}
.searchIcon{
    color: #00000040;
}
.sortBy{
    color: #00000040
}
.searchInput{
    width:100%;
    padding-left: 1%;
}
.movieDetails{
    text-align: left;
}
.default{
    min-height: 60%;
    align-items: center;
    justify-content: center;
    display: flex;
    height: auto;
}