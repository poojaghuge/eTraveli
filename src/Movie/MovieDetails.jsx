import React, { useState, useEffect } from "react";
import { Box, Grid, Paper } from "@mui/material";

const MovieDetails = (props) => {
  const { selected } = props;
  if (!selected) {
    return (
      <div className="default">
        No Movie Selected
      </div>
    );
  }

  return (
    <div className="movieDetails">
      <h4>{selected?.title}</h4>
      <section>
        <div className="description" data-testid="description">
          {selected?.opening_crawl}
        </div>
        <br />
        <div data-testid="directedBy">Directed By: {selected?.director}</div>
      </section>
    </div>
  );
};

export const MemoisedMovieDetails = React.memo(MovieDetails);
