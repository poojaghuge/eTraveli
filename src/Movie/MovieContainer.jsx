import React, { useState, useEffect, useCallback } from "react";
import { Box, Grid, Paper } from "@mui/material";
import { getMovies } from "../common/service/MovieService";
import { MemoisedBasicTable } from "../common/components/BasicTable";
import { MemoisedMovieDetails } from "./MovieDetails";
import { MemoisedActionBar } from "./ActionBar";
import { sortArray } from "../common/utility/utilityMethods";

export const MovieContainer = () => {
  const [response, setResponse] = useState([]);
  const [data, setData] = useState([]);
  const [selected, setSelected] = useState();
  const [loading, setLoading] = useState(true);

  const getData = async () => {
    const response = await getMovies();
    setResponse(response.results);
    setData(response.results);
  };

  const showDetails = (item) => {
    setSelected(item);
  };

  const handleChange = (val) => {
    if (!val) {
      setData(response);
      return;
    }
    const result = data.filter((item) => {
      return item.title.toLowerCase().includes(val.toLowerCase());
    });
    setData(result);
  };

  const handleSort = (order) => {
    console.log(order);
    const array = sortArray([...data], order);
    console.log(array);
    setData(array);
  };

  useEffect(() => {
    getData();
    setLoading(false);
  }, []);

  const showDetailsMemo = useCallback(showDetails, []);
  const handleSortMemo = useCallback(handleSort, [data]);

  return (
    <>
      <MemoisedActionBar
        handleChange={handleChange}
        handleSort={handleSortMemo}
      />
      {loading ? (
        <div>Loading...</div>
      ) : (
        <Grid
          container
          rowSpacing={1}
          columnSpacing={{ xs: 1, sm: 2, md: 3 }}
          className="container">
          <Grid item xs={7}>
            <MemoisedBasicTable data={data} showDetails={showDetailsMemo} />
          </Grid>
          <Grid item xs={5}>
            <MemoisedMovieDetails selected={selected} />
          </Grid>
        </Grid>
      )}
    </>
  );
};
