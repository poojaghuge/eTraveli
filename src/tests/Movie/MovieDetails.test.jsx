import { render, screen } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import {MemoisedMovieDetails} from "../../Movie/MovieDetails";

const item = {
    "title": "The Phantom Menace",
    "episode_id": 1,
    "opening_crawl": "Turmoil has engulfed the\r\nGalactic Republic. The taxation\r\nof trade routes to outlying star\r\nsystems is in dispute.\r\n\r\nHoping to resolve the matter\r\nwith a blockade of deadly\r\nbattleships, the greedy Trade\r\nFederation has stopped all\r\nshipping to the small planet\r\nof Naboo.\r\n\r\nWhile the Congress of the\r\nRepublic endlessly debates\r\nthis alarming chain of events,\r\nthe Supreme Chancellor has\r\nsecretly dispatched two Jedi\r\nKnights, the guardians of\r\npeace and justice in the\r\ngalaxy, to settle the conflict....",
    "director": "George Lucas",
    "producer": "Rick McCallum",
    "release_date": "1999-05-19",
    "characters": [],
    "planets": [ ],
    "starships": [],
    "vehicles": [],
    "species": [],
    "created": "2014-12-19T16:52:55.740000Z",
    "edited": "2014-12-20T10:54:07.216000Z",
    "url": "https://swapi.dev/api/films/4/"
}

describe("Test MemoisedMovieDetails", () => {
  it("test if matches the DOM Snapshot", () => {
    const domTree = renderer.create(<MemoisedMovieDetails />).toJSON();
    expect(domTree).toMatchSnapshot();
  });

   it("test if selected prop is not passed default div should render", () => {
    const {getByText} = render(<MemoisedMovieDetails selected={null}/>);
    const element = getByText('No Movie Selected');
    expect(element).toBeInTheDocument();
  });

   it("test if selected prop is passed then movie detail view should render", () => {
    const {getByText} = render(<MemoisedMovieDetails selected={item}/>);

    const Title = getByText("The Phantom Menace");
    expect(Title).toBeInTheDocument();
    expect(screen.getByTestId('description').textContent).toBe(item.opening_crawl)
    expect(screen.getByTestId('directedBy').textContent).toBe('Directed By: ' + item.director)
  });
});
