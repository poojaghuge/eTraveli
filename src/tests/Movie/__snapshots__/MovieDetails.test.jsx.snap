// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MemoisedMovieDetails test if matches the DOM Snapshot 1`] = `
<div
  className="default"
>
  No Movie Selected
</div>
`;
